### ISSUE TITLE
Defer level and level asset loading until level first activated

### ISSUE THREAD

**Issue by collinhover** [collinhover/impactplusplus#32]
@Vassildador brought up the idea of a deferred loading of level js file and assets until the level is first activated in #31. This would cut down initial load time greatly. Potentially instead of loading the next level when it is activated, we could search the level for level change trigger and preload that level? If none is found, we could either do nothing or preload all remaining levels while the player plays the first level?

**Comment by bdaenen** [comment]
Been working on this quite a lot today and have tried different approaches. Currently I've resorted to preloading all levels referenced to from the current level by level change entities like you suggested. I've made it manually accessible as well and it takes the name of the level as an argument so it can easily be used in any way one would like. It needs some cleaning up and handling for loading multiple levels in parallel but I'm getting there! :)

I created an extra helper to handle XHR requests but have only implemented the necessary functionality for this feature (GET requests without any additional parameters and the crafting of a cross-browser compatible object which will is re-used to avoid the need for garbage collection).

Any ideas for slow connections where a level could possibly not have finished loading by the time we try to load it :)? I'm currently keeping track if I'm preloading something or not, but can't think of a clean way to deal with actually delaying the loading of the level.

Off-topic: since updating impact++ dev-branch the camera doesn't seem to work anymore after changing levels in a top-down game. On top of that my Player doesn't spawn on the Checkpoint I'm using as the playerSpawner, but on the position I placed it in Weltmeister. Any idea what could be causing this?

**Comment by collinhover** [comment]
Welcome back :-)

What if we used a loader as a transition between levels when the next is still loading? Basically the same as the loader we use to preload the game, but it has slightly different functionality and a default logo that says something like "loading next level...". It could manage the preloading in the background while the current level is being played. Would that work with what you have?

Off topic, I'm reworking triggers a bit to make sure when chaining triggers that each trigger's delay (and some other properties) are taken into account. I also moved the player respawning and camera handling after death into the checkpoint, and I may have broken these things in the process. I'll look into the issues you mentioned, thanks for finding them.

**Comment by collinhover** [comment]
Camera and player spawning issues should be fixed in https://github.com/collinhover/impactplusplus/commit/768947d2b747bb81dfebfac2eefa321f8c95c3c8

**Comment by bdaenen** [comment]
I've been refactoring and improving on this quite a bit today. Currently there is an `ig.Preloader` which is created on the `ig.gameExtended` object. When building the level I keep track of `levelChange` entities and add their `settings.levelName` properties into the preloader's queue. After that I call my preloader's `loadLevels()` which will load the <levelname>.js files, extract the images used in that level, check if they're already loaded and make new Images out of them if they're not. When all images are loaded (gotta love signals.js) the preloader's `ready` is set to `true` and level loading will happen as before.

When attempting to change levels while the preloader's `ready` is still `false`, however, the preloader will be passed the name/object of the level that attempted to load. An event `onReady` will be bound and loading will stop. When the event is dispatched `ig.game.loadLevelDeferred()` will be called and passed the object of the level that we caught earlier.

The only issue I'm running into is actually showing a loader or something similar. I simply can't seem to get it to draw >.<. On top of that, I think pausing the game during this time is a necessity as well, so calling `ig.game.pause()` provides an (imo) proper solution. The only thing to change would be the text printed on the overlay :)

What do you think :)?

**Comment by collinhover** [comment]
Wow, very nice. So far I think I'd have set everything up the same way. In terms of pausing, I would actually unload the current level even if the next is not loaded. That avoids the issue of pausing the game, because nothing will be active. Currently the game uses a ui overlay to fade the current level out and the new level in... can the preloader extend the ui overlay and replace the game's transitioner?

**Comment by bdaenen** [comment]
I'll look into it, that shouldn't be too big of a problem and looks like a nice way to implement it :)
